jumanji/environments/logic/sliding_tile_puzzle/env.py

```diff
@@ -0,0 +1,255 @@
+                Defaults to `RandomGenerator` which generates puzzles with
+                a size of 5x5.
+
+            viewer: environment viewer for rendering.
+        """
+        n = 5
+        # The generator generates a shuffled puzzle
+        self.generator = generator or RandomGenerator(n)
+
+        # Create viewer used for rendering
+        self._env_viewer = viewer or SlidingTilePuzzleViewer(name="SlidingTilePuzzle")
+        self.movements = jnp.array(
+            [[-1, 0], [1, 0], [0, -1], [0, 1]]  # Up  # Down  # Left  # Right
+        )
+        self.solved_puzzle = jnp.arange(n**2).reshape((n, n))
+
+    def reset(self, key: chex.PRNGKey) -> Tuple[State, TimeStep[Observation]]:
+        """Resets the environment to an initial state."""
+        key, subkey = jax.random.split(key)
+        puzzle, empty_tile_position = self.generator(subkey)
+        state = State(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            key=key,
+        )
+        action_mask = self._get_valid_actions(empty_tile_position)
+        obs = Observation(
+            puzzle=puzzle,
+            empty_tile_position=empty_tile_position,
+            action_mask=action_mask,
+        )
+        timestep = restart(observation=obs)
+        return state, timestep
+
+    def step(
+        self, state: State, action: chex.Array
+    ) -> Tuple[State, TimeStep[Observation]]:
+        """Updates the environment state after the agent takes an action."""
+        (updated_puzzle, updated_empty_tile_position) = self._move_empty_tile(
+            state.puzzle, state.empty_tile_position, action
+        )
+        # Check if the puzzle is solved
+        done = jnp.array_equal(updated_puzzle, self.solved_puzzle)
+
+        # Update the action mask
+        action_mask = self._get_valid_actions(updated_empty_tile_position)
+
+        next_state = State(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            key=state.key,
+        )
+        obs = Observation(
+            puzzle=updated_puzzle,
+            empty_tile_position=updated_empty_tile_position,
+            action_mask=action_mask,
+        )
+
+        reward = -jnp.sum(obs.puzzle != self.solved_puzzle).astype(jnp.float32)
+
+        timestep = lax.cond(done, termination, transition, reward, obs)
+
+        return next_state, timestep
+
+    def _move_empty_tile(
+        self,
+        puzzle: chex.Array,
+        empty_tile_position: chex.Array,
+        action: chex.Array,
+    ) -> Tuple[chex.Array, chex.Array]:
+        """Moves the empty tile in the given direction and returns the updated puzzle and reward."""
+
+        # Compute the new position
+        new_empty_tile_position = empty_tile_position + self.movements[action]
+
+        # Predicate for the conditional
+        is_valid_move = jnp.all(
+            (new_empty_tile_position >= 0) & (new_empty_tile_position < puzzle.shape[0])
```

Comment:
Instead of `new_empty_tile_position < puzzle.shape[0]` can you do `new_empty_tile_position < self.generator._grid_size` think it's a bit more clear.

Note: might need to be `_grid_size - 1`, not quite sure